pipeline-model-definition/src/test/resources/environmentCrossReferences.groovy

```diff
@@ -0,0 +1,52 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+pipeline {
+    environment {
+        FOO = "FOO"
+        BAR = '${FOO}BAR'
+    }
+    agent {
+        label "some-label"
+    }
+
+    stages {
+        stage("foo") {
+            environment {
+                BAZ = "${FOO}BAZ"
+                SPLODE = "${params.WUT ?: 'banana'}"
```

Comment:
So...I should probably rename the variable from `SPLODE`, since, well, it doesn't 'splode. It works. =)

Yes, `BAR = "${BAR}FOO"` would result in echoing `BAR is FOOBARFOO`, and yes, the stage overrides would "reset" - they're only scoped in the context of the `withEnv` block within the stage they're specified in.